### DISCUSSION TITLE
SMCB-authentication: encoding of challenge

### DISCUSSION THREAD

**Discussion by mge-medisoftware** [Digitaler-Impfnachweis/certification-apis#24] [Q&A] [answered]
Im Authentication Flow steht bei "3. Perform Signature operation with SMCB" der Satz:
"The signature is created based on the provided challenge and does not need to be hashed."

Nach gematik-Spezifikation zum Konnektor unterstützt die Operation ExternalAuthenticate beim XML-Element BinaryString/Base64Date nur Strings mit bestimmten Längen (aus Länge wird auf verwendete Hashverfahren geschlossen). Es werden folgende Längen unterstützt:
- 256 Bit: SHA-256
- 384 Bit: SHA-384
- 512 Bit: SHA-512

Die x-auth-challenge wie  171c09de-2f1a-493a-bad7-ee4761a2bec3  ist eine UUID also letztlich eine 128-Bit-Zahl und kann somit nicht direkt verwendet werden!

Man kann z.B. bei der UUID die Bindestriche entfernen und dann den String(!) base64 enkodieren (das gibt dann 256 Bit).
Oder soll vorher über die UUID (mit Bindestrichen) eines der Hashverfahren angewendet werden?
Wenn ja, welches?

**Comment by oliver-steinbrecher** [comment]
Wir sind in der Prüfung

**Comment by oliver-steinbrecher** [comment, marked as answer]
Die Dokumentation war nicht korrekt formuliert. Es ist `256 Bit: SHA-256` zu verwenden.

**Reply by mge-medisoftware** [reply]
Hallo Herr Steinbrecher,

vielen Dank für die rasche Antwort, also einfach UUID-String mit SHA256 hashen 

"171c09de-2f1a-493a-bad7-ee4761a2bec3" wird z.B. derart gehasht dann base64-kodiert zu  "UiLH4ZjGKGlGhyaGfdkBgaKQ21z6j74wTiKYjtajVkg=" (das akzeptiert der Konnektor bei Operation ExternalAuthenticate tatsächlich auch)

**Reply by mge-medisoftware** [reply]
Noch ein Nachschlag: Sollte man auch genau dokumentieren, alternativ kann ja auch UUID in 128-Bit-Zahl  gewandelt werden und dann in dieser binären 16-Byte-Darstellung  mit SHA256 gehasht werden.

**Reply by mge-medisoftware** [reply]
Die Korrektur von heute "The signature is created based on the provided challenge and must be hashed with SHA-256." ist irreführend, da ja nicht die Signatur gehasht werden soll.
